tephra-core/src/main/java/org/apache/tephra/persist/AbstractTransactionLog.java

```diff
@@ -85,105 +100,145 @@ public long getTimestamp() {
+      stopWatch.stop();
+      long elapsed = stopWatch.elapsedMillis();
+      long bytesWritten = writer.getPosition() - positionBeforeWrite;
+      if (elapsed >= slowAppendThreshold) {
+        LOG.info("Slow append to log {}, took {} ms for {} entr{} and {} bytes.",
+                 getName(), elapsed, countSinceLastSync, countSinceLastSync == 1 ? "y" : "ies", bytesWritten);
+      }
+      metricsCollector.histogram("wal.sync.size", countSinceLastSync);
+      metricsCollector.histogram("wal.sync.bytes", (int) bytesWritten); // single sync won't exceed max int
     }
+    positionBeforeWrite = -1L;
+    countSinceLastSync = 0;
   }
 
   private void sync() throws IOException {
     // writes out pending entries to the HLog
-    TransactionLogWriter tmpWriter = null;
     long latestSeq = 0;
     int entryCount = 0;
     synchronized (this) {
       if (closed) {
-        return;
-      }
-      // prevent writer being dereferenced
-      tmpWriter = writer;
-
-      List<Entry> currentPending = getPendingWrites();
-      if (!currentPending.isEmpty()) {
-        tmpWriter.commitMarker(currentPending.size());
+        if (pendingWrites.isEmpty()) {
+          // this expected: close() sets closed to true after syncing all pending writes (including ours)
+          return;
+        }
+        // this should never happen because close() only sets closed=true after syncing.
+        // but if it should happen, we must fail this call because we don't know whether the edit was persisted
+        throw new IOException(
+          "Unexpected state: Writer is closed but there are pending edits. Cannot guarantee that edits were persisted");
       }
-
-      // write out all accumulated entries to log.
-      for (Entry e : currentPending) {
-        tmpWriter.append(e);
-        entryCount++;
-        latestSeq = Math.max(latestSeq, e.getKey().get());
+      Entry[] currentPending = getPendingWrites();
+      if (currentPending != null) {
+        entryCount = currentPending.length;
+        startTimerIfNeeded(writer, entryCount);
+        writer.commitMarker(entryCount);
+        for (Entry e : currentPending) {
+          writer.append(e);
+        }
+        // sequence are guaranteed to be ascending, so the last one is the greatest
+        latestSeq = currentPending[currentPending.length - 1].getKey().get();
+        writtenUpTo = latestSeq;
       }
     }
 
-    long lastSynced = syncedUpTo.get();
+    // giving up the sync lock here allows other threads to write their edits before the sync happens.
+    // hence, we can have the edits from n threads in one sync.
+
     // someone else might have already synced our edits, avoid double syncing
-    if (lastSynced < latestSeq) {
-      tmpWriter.sync();
-      metricsCollector.histogram("wal.sync.size", entryCount);
-      syncedUpTo.compareAndSet(lastSynced, latestSeq);
+    if (syncedUpTo < latestSeq) {
```

Comment:
Since this check is happening outside the synchronized block, `latestSeq` may not give the right value - we can just use `writtenUpTo` instead of `latestSeq`. 

To be safe, let's move the definition of both `latestSeq` and `entryCount` into the synchronized block on line 196, so that they cannot be used outside the block.

Reply:
Please see the previous discussion of this at https://github.com/apache/incubator-tephra/pull/53#pullrequestreview-62374698